syntax = "proto3";

option go_package = "magma/dp/cloud/go/active_mode_controller/protos/active_mode";

service ActiveModeController {
  rpc GetState(GetStateRequest) returns (State) {}
  rpc ToggleActiveMode (ToggleActiveModeParams) returns (ToggleActiveModeResponse) {}
}

message GetStateRequest {}

message State {
  repeated ActiveModeConfig active_mode_configs = 1;
}

message ActiveModeConfig {
  CbsdState desired_state = 1;
  Cbsd cbsd = 2;
}

enum CbsdState {
  Unregistered = 0;
  Registered = 1;
}

message Cbsd {
  string id = 1;
  string user_id = 2;
  string fcc_id = 3;
  string serial_number = 4;
  CbsdState state = 5;
  repeated Grant grants = 6;
  repeated Channel channels = 7;
  optional float eirp_capability = 8;
  repeated string pending_requests = 9;
}

message Grant {
  string id = 1;
  GrantState state = 2;
  int64 heartbeat_interval_sec = 3;
  int64 last_heartbeat_timestamp = 4;
}

enum GrantState {
  Granted = 0;
  Authorized = 1;
}

message Channel {
  FrequencyRange frequency_range = 1;
  optional float max_eirp = 2;
  optional float last_eirp = 3;
}

message FrequencyRange {
  int64 low = 1;
  int64 high = 2;
}

enum Switch {
  On = 0;
  Off = 1;
}

message ToggleActiveModeParams {
  int64 cbsd_id = 1;
  Switch switch = 2;
  optional CbsdState desired_state = 3;
}

message ToggleActiveModeResponse {
  int64 cbsd_db_id = 1;
  bool active_mode_enabled = 2;
  optional CbsdState cbsd_state = 3;
}
